### PR TITLE
Add alternative to Spy Nodes that can fully participate in Gossip

### PR DESCRIPTION
**Pull request by sagar-solana**
#### Problem

Trying to discover all nodes in a cluster with just Gossip Pulls is insufficient. However having our spy nodes support this is beneficial when the client is behind a NAT

#### Summary of Changes

Added a mode to solana-gossip and ClusterInfo::discover to support fully participating in gossip.

### CODE REVIEW

File: gossip/src/main.rs [hunk 1]
```diff
@@ -111,7 +123,25 @@ fn main() -> Result<(), Box<dyn error::Error>> {
                 .value_of("node_pubkey")
                 .map(|pubkey_str| pubkey_str.parse::<Pubkey>().unwrap());
 
-            let nodes = discover(&network_addr, num_nodes, timeout, pubkey)?;
+            let gossip_addr = if matches.is_present("pull_only") {
+                None
+            } else {
+                let mut addr: SocketAddr = "0.0.0.0:0000".parse().unwrap();
```
Comment (pgarg66) [inline gossip/src/main.rs]: You can use ```socketaddr_any!();```

**Comment by sagar-solana** [comment]
FYI, @CriesofCarrots. Solana-gossip now uses a full gossip node for "spy" and so it needs the machine to be visible to the cluster. The previous behavior is behind a flag, --pull-only. Just incase you see any third party running into this.

**Comment by mvines** [comment]
oh that's a good point, @sagar-solana tomorrow or something can you make a quick update to book/src/testnet-participation.md that mentions adds the new --public-address to `solana-gossip` example there and any other info that seems relevant.  Folks standing up validators will certainly need to be aware

**Comment by sagar-solana** [comment]
Will do!

**Pull request merged**